// selenium-webdriver/bidi/generated/bluetooth.js
"use strict";
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bluetooth = void 0;
// --- Implementation ---
class Bluetooth {
    constructor(bidi) {
        this.bidi = bidi;
    }
    static async create(driver) {
        const caps = await driver.getCapabilities();
        if (!caps.get('webSocketUrl')) {
            throw new Error('WebDriver instance must support BiDi protocol');
        }
        const bidi = await driver.getBidi();
        return new Bluetooth(bidi);
    }
    async handleRequestDevicePrompt(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.handleRequestDevicePrompt',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulateAdapter(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulateAdapter',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async disableSimulation(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.disableSimulation',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulatePreconnectedPeripheral(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulatePreconnectedPeripheral',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulateAdvertisement(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulateAdvertisement',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulateGattConnectionResponse(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulateGattConnectionResponse',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulateGattDisconnection(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulateGattDisconnection',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulateService(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulateService',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulateCharacteristic(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulateCharacteristic',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulateCharacteristicResponse(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulateCharacteristicResponse',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulateDescriptor(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulateDescriptor',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async simulateDescriptorResponse(params) {
        const response = await this.bidi.send({
            method: 'bluetooth.simulateDescriptorResponse',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async onRequestDevicePromptUpdated(callback) {
        await this.bidi.subscribe('bluetooth.requestDevicePromptUpdated');
        this.bidi.on('bluetooth.requestDevicePromptUpdated', (params) => {
            callback(params);
        });
    }
    async onGattConnectionAttempted(callback) {
        await this.bidi.subscribe('bluetooth.gattConnectionAttempted');
        this.bidi.on('bluetooth.gattConnectionAttempted', (params) => {
            callback(params);
        });
    }
}
exports.Bluetooth = Bluetooth;
