// selenium-webdriver/bidi/generated/emulation.js
"use strict";
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Emulation = void 0;
// --- Implementation ---
class Emulation {
    constructor(bidi) {
        this.bidi = bidi;
    }
    static async create(driver) {
        const caps = await driver.getCapabilities();
        if (!caps.get('webSocketUrl')) {
            throw new Error('WebDriver instance must support BiDi protocol');
        }
        const bidi = await driver.getBidi();
        return new Emulation(bidi);
    }
    async setForcedColorsModeThemeOverride(params) {
        const response = await this.bidi.send({
            method: 'emulation.setForcedColorsModeThemeOverride',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setGeolocationOverride(params) {
        const response = await this.bidi.send({
            method: 'emulation.setGeolocationOverride',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setLocaleOverride(params) {
        const response = await this.bidi.send({
            method: 'emulation.setLocaleOverride',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setNetworkConditions(params) {
        const response = await this.bidi.send({
            method: 'emulation.setNetworkConditions',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setScreenOrientationOverride(params) {
        const response = await this.bidi.send({
            method: 'emulation.setScreenOrientationOverride',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setScreenSettingsOverride(params) {
        const response = await this.bidi.send({
            method: 'emulation.setScreenSettingsOverride',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setScriptingEnabled(params) {
        const response = await this.bidi.send({
            method: 'emulation.setScriptingEnabled',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setScrollbarTypeOverride(params) {
        const response = await this.bidi.send({
            method: 'emulation.setScrollbarTypeOverride',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setTimezoneOverride(params) {
        const response = await this.bidi.send({
            method: 'emulation.setTimezoneOverride',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setTouchOverride(params) {
        const response = await this.bidi.send({
            method: 'emulation.setTouchOverride',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setUserAgentOverride(params) {
        const response = await this.bidi.send({
            method: 'emulation.setUserAgentOverride',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
}
exports.Emulation = Emulation;
