// selenium-webdriver/bidi/generated/input.js
"use strict";
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Input = void 0;
exports.getInputInstance = getInputInstance;
// --- Additional Types ---
/** @deprecated Use {@link Input.create} instead — will be removed in a future major version. */
async function getInputInstance(driver) {
    return Input.create(driver);
}
// --- Implementation ---
class Input {
    constructor(bidi) {
        this.bidi = bidi;
    }
    static async create(driver) {
        const caps = await driver.getCapabilities();
        if (!caps.get('webSocketUrl')) {
            throw new Error('WebDriver instance must support BiDi protocol');
        }
        const bidi = await driver.getBidi();
        return new Input(bidi);
    }
    async performActions(params) {
        const response = await this.bidi.send({
            method: 'input.performActions',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async releaseActions(params) {
        const response = await this.bidi.send({
            method: 'input.releaseActions',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async setFiles(params) {
        const response = await this.bidi.send({
            method: 'input.setFiles',
            params: params,
        });
        if (response['type'] === 'error') {
            throw new Error(`${response['error']}: ${response['message']}`);
        }
    }
    async onFileDialogOpened(callback) {
        await this.bidi.subscribe('input.fileDialogOpened');
        this.bidi.on('input.fileDialogOpened', (params) => {
            callback(params);
        });
    }
    /** @deprecated Use {@link performActions} instead — will be removed in a future major version. */
    async perform(browsingContextId, actions) {
        await this.performActions({ context: browsingContextId, actions });
    }
    /** @deprecated Use {@link releaseActions} instead — will be removed in a future major version. */
    async release(browsingContextId) {
        await this.releaseActions({ context: browsingContextId });
    }
}
exports.Input = Input;
