// selenium-webdriver/bidi/generated/log.js
"use strict";
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Log = void 0;
exports.getLogInspectorInstance = getLogInspectorInstance;
// --- Additional Types ---
/** @deprecated Use {@link Log.create} instead — will be removed in a future major version. */
async function getLogInspectorInstance(driver) {
    return Log.create(driver);
}
// --- Implementation ---
class Log {
    constructor(bidi) {
        this.bidi = bidi;
    }
    static async create(driver) {
        const caps = await driver.getCapabilities();
        if (!caps.get('webSocketUrl')) {
            throw new Error('WebDriver instance must support BiDi protocol');
        }
        const bidi = await driver.getBidi();
        return new Log(bidi);
    }
    async onEntryAdded(callback) {
        await this.bidi.subscribe('log.entryAdded');
        this.bidi.on('log.entryAdded', (params) => {
            callback(params);
        });
    }
    async onConsoleEntry(callback) {
        await this.onEntryAdded((entry) => {
            if (entry.type === 'console')
                callback(entry);
        });
    }
    async onJavascriptLog(callback) {
        await this.onEntryAdded((entry) => {
            if (entry.type === 'javascript')
                callback(entry);
        });
    }
    async onJavascriptException(callback) {
        await this.onEntryAdded((entry) => {
            if (entry.type === 'javascript' && entry.level === 'error')
                callback(entry);
        });
    }
    async onLog(callback) {
        await this.onEntryAdded(callback);
    }
}
exports.Log = Log;
